### ISSUE TITLE
BlenderBIM : IfcRelSpaceBoundaries capabilities

### ISSUE THREAD

**Issue by CyrilWaechter** [IfcOpenShell/IfcOpenShell#1676]
It would be great to enhance IfcRelSpaceBoundary capabilities inside BlenderBIM Add-on with following:

- [ ] Display geometry
- [ ] Display attributes
- [ ] Modify geometry
- [ ] Modify attributes
- [ ] Apply consistent colors based on related building element similarly to [ladybug gbXML viewer](https://www.ladybug.tools/spider/gbxml-viewer/r12/gv-app/gv-app.html)
- [ ] Build test suite to prevent common mistakes or warn about common issues coming from other authoring tools (bimtester report ?)
- [ ] Tools to fix common mistakes from software vendors (ifcpatch + tools to help enhancing attributes data)

It would allow to use BlenderBIM Add-on as:

- Software agnostic IFC energy model auditing tool
- Basic IFC energy model creation tool
- IFC energy model improvement tool

**Comment by Moult** [comment]
Have you got some sample files? This will help dev.

**Comment by CyrilWaechter** [comment]
Yes multiple sample files have been produced during BIMxBEM project: [IfcTestFiles](https://github.com/ENAC-CNPA/BIMxBEM/tree/master/IfcTestFiles)
There is also Aalseth Lane Residence somewhere else.

**Comment by brunopostle** [comment]
@CyrilWaechter since each ifcrelspaceboundary has to be attached to exactly one ifcspace, these external surfaces of external walls don't seem to be supported in IFC (unless we create an ifcspace representing the entire outside world, but with no Representation)

**Comment by CyrilWaechter** [comment]
@brunopostle Hmmm, interesting. We need in some way identify what is `EXTERNAL`, `EXTERNAL_EARTH`, `EXTERNAL_WATER` and `EXTERNAL_FIRE`. Having external spaces might simplify this process and some others. The same way you don't represent your ground with infinite depth external spaces could be defined inside an extended bounding box of the project.

External surfaces of external walls are not in the IFC specifications anyway so nothing prevent you to create another kind of IfcRelSpaceBoundary (not 1st level or 2nd level) to represent it. Thermal breaks are also not defined in IFC so you also need to use an hacky way to store it but this is another story.

**Comment by aothms** [comment]
I find it quite a nice approach to have a space with spacetypenum .EXTERNAL. https://standards.buildingsmart.org/IFC/DEV/IFC4_3/RC2/HTML/schema/ifcproductextension/lexical/ifcspacetypeenum.htm but one single external space can have weird consequences for e.g spatial connectivity graphs, because it would basically connect to every external door, but doesn't mean you can actually reach it through that space, e.g balcony or inner courtyards. But that's a typical difficulty I guess with multi-disciplinary models. A common thing like space doesn't generalize that well across disciplines.

**Comment by brunopostle** [comment]
I have some more confusion regarding space boundaries :(

I seem to be able to create 2nd level boundaries for windows, and can use the **ParentBoundary** attribute to relate them to another boundary representing a wall.

This works fine unless I have two windows in a wall, like so:

```
#667=IFCRELSPACEBOUNDARY2NDLEVEL('1$zWU3paDAOOqGDH_YXN4t',#666,$,'CellIndex 0',#1157,#586,#668,.PHYSICAL.,.EXTERNAL.,$,$);
#727=IFCRELSPACEBOUNDARY2NDLEVEL('242$Va5K59KA4Njri0_S1M',#726,$,'CellIndex 0',#1157,#693,#728,.PHYSICAL.,.EXTERNAL.,#667,$);
#779=IFCRELSPACEBOUNDARY2NDLEVEL('3SY$ocyKnBwOGPHA5QuH8t',#778,$,'CellIndex 0',#1157,#745,#780,.PHYSICAL.,.EXTERNAL.,#667,$);
```

`ifcopenshell.validate` complains that I have more than one boundary in the **Corresponds** attribute of the parent, but I haven't set any **CorrespondingBoundary** attributes?

```
In #667=IfcRelSpaceBoundary2ndLevel('1$zWU3paDAOOqGDH_YXN4t',#666,$,'CellIndex 0',#1157,#586,#668,.PHYSICAL.,.EXTERNAL.,$,$)
(#727=IfcRelSpaceBoundary2ndLevel('242$Va5K59KA4Njri0_S1M',#726,$,'CellIndex 0',#1157,#693,#728,.PHYSICAL.,.EXTERNAL.,#667,$), #779=IfcRelSpaceBoundary2ndLevel('3SY$ocyKnBwOGPHA5QuH8t',#778,$,'CellIndex 0',#1157,#745,#780,.PHYSICAL.,.EXTERNAL.,#667,$)) not valid for <inverse Corresponds: set [0:1] of <entity IfcRelSpaceBoundary2ndLevel> for <attribute CorrespondingBoundary?: <entity IfcRelSpaceBoundary2ndLevel>>>
``` 

Here's a file with four walls and six windows that fails to validate: [tmpm097so2c.ifc.gz](https://github.com/IfcOpenShell/IfcOpenShell/files/7108853/tmpm097so2c.ifc.gz)

**Comment by aothms** [comment]
Wow you stumbled on a rather elementary bug. I guess it must be rare in the schema to have inverses for optional attributes... The commit fixes this.

**Comment by Moult** [comment]
Just curious what the status is on this bug and what we can do to progress it? @CyrilWaechter

**Comment by CyrilWaechter** [comment]
> Just curious what the status is on this bug and what we can do to progress it? @CyrilWaechter

Initial comment is up to date. I still need to develop a test suite, tools to fix common mistakes and find a way to have a energy specific view with a one click button.

**Comment by CyrilWaechter** [comment]
@Moult Following our call I have updated the first post with discussed topics.
